test/run_tests.pl

```diff
@@ -82,15 +92,15 @@ BEGIN
     if ($arg eq 'alltests') {
 	warn "'alltests' encountered, ignoring everything before that...\n"
 	    unless $initial_arg;
-	%tests = map { $_ => basename($_) } @alltests;
+	%tests = map { reorder($_) => $_ } @alltests;
```

Comment:
This isn't right. Look for `sort keys %tests`, change it to `sort { ... } keys %tests`, where '...' is your sorting code.

Reply:
No, it does exactly what I meant it to do:
`map { reorder($_) => $_ } @alltests` transforms an array of test script names into a hash table
where each element associates with the string `reorder($_)` the original name `$_`.

The `=>` may look confusing here because it does not mean a lamba abstraction that typically may be used to define a function on-the-fly.

Note that the `reorder` subroutine just 'renames' some of the test scritps, and together with the construct explained above it effectively re-orders the names where needed.

I had already considered using the optional `{ ... }` argument with `sort`, but this has a different semantics: compare two values and yield -1, 0, or +1.

Reply:
I have no idea why I'd interpret the list pair operator (that's really what `=>` is) as a lambda abstraction.  Mixed up languages, perhaps?

I don't quite understand your retort,  The "reorder" function simply rewrites the keys to get a different *sorting order*, so why that would be impossible to use *when sorting*, I do not understand.  The following (diff from master) should give the exact same results.

``` diff
diff --git a/test/run_tests.pl b/test/run_tests.pl
index d47f7cf1e6..864bc7a5b4 100644
--- a/test/run_tests.pl
+++ b/test/run_tests.pl
@@ -70,6 +70,16 @@ open $openssl_args{'tap_copy'}, ">$outfilename"
 my @alltests = find_matching_tests("*");
 my %tests = ();
 
+sub reorder {
+    my $key = pop;
+
+    # for parallel test runs, do slow tests first
+    if (defined $jobs && $jobs > 1 && $key =~ m/test_ssl_new|test_fuzz/) {
+        $key =~ s/(\d+)-/00-/;
+    }
+    return $key;
+}
+
 my $initial_arg = 1;
 foreach my $arg (@ARGV ? @ARGV : ('alltests')) {
     if ($arg eq 'list') {
@@ -281,7 +291,7 @@ unless (defined $eres) {
 my $harness = $package->new(\%tapargs);
 my $ret =
     $harness->runtests(map { [ abs2rel($_, rel2abs(curdir())), $tests{$_} ] }
-                       sort keys %tests);
+                       sort { reorder($a) cmp reorder($b) } keys %tests);
 
 # $ret->has_errors may be any number, not just 0 or 1.  On VMS, numbers
 # from 2 and on are used as is as VMS statuses, which has severity encoded
```

Reply:
> I have no idea why I'd interpret the list pair operator (that's really what `=>` is) as a lambda abstraction. Mixed up languages, perhaps?

At least to me the `=>` is confusing, from the background of functional languages or the lambda calculus underlying them.

> I don't quite understand your retort, The `reorder` function simply rewrites the keys to get a different _sorting order_, 

That's correct. My "no" was just on your strict "This isn't right.", which I don't agree on.

> so why that would be impossible to use _when sorting_, I do not understand.

I did not see that it was possible also via the `{ ...}` argument to `sort`.

> The following (diff from master) should give the exact same results.
> ```
>      $harness->runtests(map { [ abs2rel($_, rel2abs(curdir())), $tests{$_} ] }
> -                       sort keys %tests);
> +                       sort { reorder($a) cmp reorder($b) } keys %tests);
>  
>  # $ret->has_errors may be any number, not just 0 or 1.  On VMS, numbers
>  # from 2 and on are used as is as VMS statuses, which has severity encoded
> ```

You are right, this works as well - thanks for pointing this out.
And it's likely easier to understand, so I've taken over your solution.

Still I've kept moving the hitherto three calls of `basename()` to one place,
namely into the `map { ... }` argument just above calling `sort`.
This is a simplification of the original code (and is independent of the re-ordering).

Reply:
> And it's likely easier to understand

That is precisely the reason why I suggested this.  I'm sorry that it came out harsher.

> Still I've kept moving the hitherto three calls of `basename()` to one place,

I suspected you would.  That's fine :wink:

> At least to me the `=>` is confusing, from the background of functional languages or the lambda calculus underlying them.

I know, I've seen it in Standard ML and in Javascript.  That's why I asked if you got languages mixed up.

But yeah, in perl, it's just syntactic sugar for the comma operator in a list context.

Side note: Out of interest, I had a look through https://en.wikipedia.org/wiki/Anonymous_function, and it turns that `->` is more common than `=>` for a lambda constructor (even in C++11, which is *terribly* confusing)

Reply:
Thanks @levitte for clarifying!